src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadBufferState.cs

```diff
@@ -2,32 +2,168 @@
+#if BUILDING_INBOX_LIBRARY
+                    _buffer.AsSpan(_count));
+#else
+                    _buffer, _count, _buffer.Length - _count);
+#endif
+
+                if (bytesRead == 0)
+                {
+                    _isFinalBlock = true;
+                    break;
+                }
+
+                _count += bytesRead;
+            }
+            while (_count < _buffer.Length);
+
+            ProcessReadBytes();
+        }
+
+        /// <summary>
+        /// Advances the buffer in anticipation of a subsequent read operation.
+        /// </summary>
+        public void AdvanceBuffer(int bytesConsumed)
+        {
+            Debug.Assert(bytesConsumed <= _count);
+            Debug.Assert(!_isFinalBlock || _count == bytesConsumed, "The reader should have thrown if we have remaining bytes.");
+
+            _count -= bytesConsumed;
+
+            if (!_isFinalBlock)
+            {
+                // Check if we need to shift or expand the buffer because there wasn't enough data to complete deserialization.
+                if ((uint)_count > ((uint)_buffer.Length / 2))
+                {
+                    // We have less than half the buffer available, double the buffer size.
+                    byte[] oldBuffer = _buffer;
+                    int oldMaxCount = _maxCount;
+                    byte[] newBuffer = ArrayPool<byte>.Shared.Rent((_buffer.Length < (int.MaxValue / 2)) ? _buffer.Length * 2 : int.MaxValue);
+
+                    // Copy the unprocessed data to the new buffer while shifting the processed bytes.
+                    Buffer.BlockCopy(oldBuffer, _offset + bytesConsumed, newBuffer, 0, _count);
+                    _buffer = newBuffer;
+                    _offset = 0;
+                    _maxCount = _count;
+
+                    // Clear and return the old buffer
+                    new Span<byte>(oldBuffer, 0, oldMaxCount).Clear();
+                    ArrayPool<byte>.Shared.Return(oldBuffer);
+                }
+                else if (_count != 0)
+                {
+                    // Shift the processed bytes to the beginning of buffer to make more room.
+                    Buffer.BlockCopy(_buffer, _offset + bytesConsumed, _buffer, 0, _count);
+                    _offset = 0;
+                }
+            }
+        }
+
+        private void ProcessReadBytes()
+        {
+            if (_count > _maxCount)
+            {
+                _maxCount = _count;
+            }
+
+            if (_isFirstBlock)
+            {
+                _isFirstBlock = false;
+
+                // Handle the UTF-8 BOM if present
+                Debug.Assert(_buffer.Length >= JsonConstants.Utf8Bom.Length);
+                if (_buffer.AsSpan(0, _count).StartsWith(JsonConstants.Utf8Bom))
+                {
+                    _offset = (byte)JsonConstants.Utf8Bom.Length;
+                    _count -= JsonConstants.Utf8Bom.Length;
+                }
+            }
```

Comment:
rather than keeping variable and special code, would it make sense to simply handle this separately (i.e. ctor or separate call in the beginning)

Reply:
Not sure I follow, this handling needs to happen after the first read.

Reply:
I was thinking about separate read but I don't really have strong preference or data telling if one way or the other is faster